# JLab-ATACFlow #


## RNA Sequencing ##

RNA Sequencing amplifies cDNA derived from poly-Adenylated mRNA to understand the transcriptome within cells.

## This pipeline

This pipeline was developed by Michael J. Bale, MSc at Weill Cornell Medical College for principal use in processing BULK RNAseq data generated by the lab of Steven Z. Josefowicz, PhD. Although some attempts at generalizing for wider use was made, some extra hacking might be required to get this pipeline to run on your specific machine and/or server. Use at your own risk; but I'm happy to help if I can!

=================================================


      R N A   s e q   P I P E L I N E v0.5
      
=================================================

Author: Michael J. Bale (mib4004@med.cornell.edu)

The typical command for running the pipeline is as follows:

`nextflow run michaelbale/JLab-RNAFlow --input 'data/*_R{1,2}.fastq.gz' --genome mouse -profile singularity`

Mandatory arguments:
```
--input						  Path to input data (must be surrounded with quotes)

--genome					  Name of Genomes reference (current supported: mouse10 -- mm10, mouse39 -- mm39, human -- hg38)
							  For mouse, unless beginning new project, likely better to keep to mm10 (03/04/2022).

-profile                      Name of package manager system (available: docker, singularity, conda);
                              for WCM default -- singularity is recommended, but conda works while docker 
                              does not. For minimal - use conda.

--boot						  Number of inferential replicates for Kallisto to run; default: 100

Options:

--executorConfig              Path to config file that contains specifics for execution. 
                              Will default to WCM SCU-specific parameters. N.B. for
                              single-threaded running use --executorConfig conf/minimal.config

--catLanes                    Tells CnRFlow to take input files and concatenate lanes into single fastq files

--name                        Project Name; cannot have whitespace characters

--workDir                     Name of folder to output concatenated fastq files to (not used unless --catLanes)

--outdir                      Name of folder to output all results to (Default: results)

--genomeAssets                Home directory of where genome-specific files are. 
                              Defaults to /athena/josefowiczlab/scratch/szj2001/JLab-Flow_Genomes
```

## Citations


## Requirements:

nextflow v20 or higher

java 8+

Some combination of singularity, docker, or conda. If using -profile debug, all programs specified in the environment.yml file will have to be installed and accessible to your path.


# JLab-RNAFlow
